cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java

```diff
@@ -923,6 +927,142 @@ public StopProposalResult stopProposalExecution(StopProposalParameters parameter
     return new StopProposalResult(_config);
   }
 
+  private void ensureNoOfflineReplicaForPartition(PartitionInfo partitionInfo, List<Node> aliveNodes) {
+    for (Node node: partitionInfo.replicas()) {
+      if (!aliveNodes.contains(node)) {
+        throw new RuntimeException(String.format("Topic partition %s-%d have offline replica on broker %d , unable to update "
+                                   + "its replication factor.", partitionInfo.topic(), partitionInfo.partition(), node.id()));
+      }
+    }
+  }
+
+  /**
+   * Update the replication factor of Kafka topics.
+   * If partition's current replication factor is less than target replication factor, add new replicas to the partition
+   * in a rack-aware, round-robin way.
+   * There are two scenarios that rack awareness property is not guaranteed.
+   * <ul>
+   *   <li> If metadata does not have rack information about brokers, then it is only guaranteed that new replicas are
+   *   added to brokers which currently do not host any replicas of partition.</li>
+   *   <li> If replication factor to set for the topic is larger than number of racks in the cluster and
+   *   skipTopicRackAwarenessCheck is set to true, then rack awareness property is ignored.</li>
+   * </ul>
+   * If partition's current replication factor is larger than target replication factor, remove one or more follower replicas
+   * from the partition.
+   *
+   * @param cluster The metadata of the cluster.
+   * @param replicationFactor The replication factor to set for the topics.
+   * @param topics The topics to apply the change.
+   * @param skipTopicRackAwarenessCheck Whether skip ignore rack awareness property if number of rack in clsuter is less
+   *                                    than target replication factor.
+   * @return Execution proposals to increase replication factor of topics.
+   */
+  private List<ExecutionProposal> maybeUpdateTopicReplicationFactor(Cluster cluster,
+                                                                    int replicationFactor,
+                                                                    Set<String> topics,
+                                                                    boolean skipTopicRackAwarenessCheck) {
+    Map<String, List<Integer>> brokersByRack = new HashMap<>();
+    Map<Integer, String> rackByBroker = new HashMap<>();
+    for (Node node : cluster.nodes()) {
+      // If the rack is not specified, we use the broker id info as rack info.
+      String rack = node.rack() == null || node.rack().isEmpty() ? String.valueOf(node.id()) : node.rack();
+      brokersByRack.putIfAbsent(rack, new ArrayList<>());
+      brokersByRack.get(rack).add(node.id());
+      rackByBroker.put(node.id(), rack);
+    }
+
+    if (replicationFactor > brokersByRack.size()) {
+      if (skipTopicRackAwarenessCheck) {
+        LOG.info(String.format("Target replication factor for topics %s is larger than number of racks in cluster, new replica"
```

Comment:
Isn't it possible that the topic replication factor is requested to be decreased -- rather than increased. Should this message be updated to be more generic -- i.e. can we avoid implying that `new replica` will be added?